src/combinatorics/composition.lean

```diff
@@ -381,31 +378,213 @@ begin
+    rw [A, this, ones_blocks] },
+end
+
+lemma ne_ones_iff {c : composition n} :
+  c ≠ ones n ↔ ∃ i ∈ c.blocks, 1 < i :=
+begin
+  rw ← not_iff_not,
+  push_neg,
+  rw eq_ones_iff,
+  have : ∀ j ∈ c.blocks, j = 1 ↔ j ≤ 1 := λ j hj, by simp [le_antisymm_iff, c.one_le_blocks hj],
+  simp [this] {contextual := tt}
+end
+
+/-- The composition made of a single block of size `n`. -/
+def single (n : ℕ) (h : 0 < n) : composition n :=
+⟨[n], by simp [h], by simp⟩
+
+@[simp] lemma single_length {n : ℕ} (h : 0 < n) : (single n h).length = 1 :=
+by simp [single, length]
+
+@[simp] lemma single_blocks {n : ℕ} (h : 0 < n) : (single n h).blocks = [n] :=
+by simp [blocks, single]
+
+@[simp] lemma single_blocks_fun {n : ℕ} (h : 0 < n) (i : fin (single n h).length) :
+  (single n h).blocks_fun i = n :=
+by simp [blocks_fun, single, blocks, i.2]
+
+@[simp] lemma single_embedding {n : ℕ} (h : 0 < n) (i : fin n) :
+  (single n h).embedding ⟨0, single_length h ▸ zero_lt_one⟩ i = i :=
+by simp [embedding]
+
+lemma eq_single_iff {n : ℕ} {h : 0 < n} {c : composition n } :
+  c = single n h ↔ c.length = 1 :=
+begin
+  split,
+  { assume H,
+    rw H,
+    exact single_length h },
+  { assume H,
+    ext1,
+    have A : c.blocks.length = 1 := H ▸ c.blocks_length,
+    have B : c.blocks.sum = n := c.blocks_sum,
+    rw eq_cons_of_length_one A at B ⊢,
+    simpa [single_blocks] using B }
 end
 
 end composition
 
+/-!
+### Splitting a list
+
+Given a list of length `n` and a composition `c` of `n`, one can split `l` into `c.length` sublists
+of respective lengths `c.blocks_fun 0`, ..., `c.blocks_fun (c.length-1)`. This is inverse to the
+join operation.
+-/
+namespace list
+variable {α : Type*}
+
+/-- Given a list of length `n` and a composition `[i₁, ..., iₖ]` of `n`, split `l` into a list of
+`k` lists corresponding to the blocks of the composition, of respective lengths `i₁`, ..., `iₖ`.
+This makes sense mostly when `n = l.length`, but this is not necessary for the definition. -/
+def split_wrt_composition (l : list α) (c : composition n) : list (list α) :=
+of_fn (λ (i : fin c.length), (l.take (c.size_up_to (i.val+1))).drop (c.size_up_to i.val))
+
+/-- When one splits a list along a composition `c`, the number of sublists thus created is
+`c.length`. -/
+@[simp] lemma length_split_wrt_composition (l : list α) (c : composition n) :
+  length (l.split_wrt_composition c) = c.length :=
+by simp [split_wrt_composition]
+
+/-- When one splits a list along a composition `c`, the lengths of the sublists thus created are
+given by the block sizes in `c`. -/
+lemma map_length_split_wrt_composition (l : list α) (c : composition l.length) :
+  map length (l.split_wrt_composition c) = c.blocks :=
+begin
+  refine ext_le (by simp) (λ i h₁ h₂, _),
+  simp [split_wrt_composition, composition.size_up_to_le c, min_eq_left, nth_le_of_fn',
+             map_of_fn, function.comp_app, length_drop, length_take],
+  rw c.blocks_length at h₂,
```

Comment:
```suggestion
```
This proof seems to work without this `rw`, but I couldn't combine the `simp` and `simp only`. The indentation looks funny too - I guess there was a `simp only` previously?